File: golang/exchange.go
package main

import (
	"math/rand"
	"time"
)

type Tick struct {
	Symbol string
	Price  float64
	Ts     time.Time
}

type Order struct {
	Symbol string
	Side   string // "BUY" or "SELL"
	Qty    float64
}

type Fill struct {
	Order     Order
	FillPrice float64
	Ts        time.Time
}

// Exchange interface.
// To swap for a real venue (Binance, Alpaca, etc.):
//   - Ticks(): spawn a goroutine that reads a WebSocket feed and pushes ticks on the channel.
//   - SubmitOrder(): call the venue's REST order endpoint and return the fill.
type Exchange interface {
	Ticks() <-chan Tick
	SubmitOrder(Order) Fill
	Stop()
}

// MockExchange: random-walk feed, instant fills at latest price.
type MockExchange struct {
	symbol   string
	price    float64
	interval time.Duration
	ch       chan Tick
	done     chan struct{}
}

func NewMockExchange(symbol string, basePrice float64, intervalMs int) *MockExchange {
	m := &MockExchange{
		symbol:   symbol,
		price:    basePrice,
		interval: time.Duration(intervalMs) * time.Millisecond,
		ch:       make(chan Tick, 64),
		done:     make(chan struct{}),
	}
	go m.run()
	return m
}

func (m *MockExchange) run() {
	t := time.NewTicker(m.interval)
	defer t.Stop()
	for {
		select {
		case <-m.done:
			close(m.ch)
			return
		case <-t.C:
			// Random walk: price moves +/- 0.1% per tick.
			m.price += (rand.Float64() - 0.5) * 0.002 * m.price
			m.ch <- Tick{Symbol: m.symbol, Price: m.price, Ts: time.Now()}
		}
	}
}

func (m *MockExchange) Ticks() <-chan Tick { return m.ch }

func (m *MockExchange) Stop() {
	select {
	case <-m.done: // already closed
	default:
		close(m.done)
	}
}

// Mock fill at current price, zero slippage, zero latency.
func (m *MockExchange) SubmitOrder(o Order) Fill {
	return Fill{Order: o, FillPrice: m.price, Ts: time.Now()}
}

File: rust/src/exchange.rs
use rand::Rng;
use std::sync::{Arc, Mutex};
use std::time::Instant;
use tokio::sync::mpsc;
use tokio::time::{interval, Duration};

#[derive(Debug, Clone)]
pub struct Tick {
    pub symbol: String,
    pub price: f64,
    pub ts: Instant,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Side {
    Buy,
    Sell,
}

#[derive(Debug, Clone)]
pub struct Order {
    pub symbol: String,
    pub side: Side,
    pub qty: f64,
}

#[derive(Debug)]
pub struct Fill {
    pub order: Order,
    pub fill_price: f64,
    pub ts: Instant,
}

/// MockExchange: random-walk feed + mock fills at latest price.
///
/// To swap for a real venue (Binance, Alpaca, Coinbase, ...):
///   - Replace this struct with one backed by `tokio-tungstenite` (WebSocket)
///     for the market-data feed and `reqwest` (REST) for order submission.
///   - Keep the same public API: `start()` returns `mpsc::Receiver<Tick>`,
///     `submit_order()` returns `Fill`.
pub struct MockExchange {
    symbol: String,
    price: Arc<Mutex<f64>>,
    interval_ms: u64,
}

impl MockExchange {
    pub fn new(symbol: &str, base_price: f64, interval_ms: u64) -> Self {
        Self {
            symbol: symbol.to_string(),
            price: Arc::new(Mutex::new(base_price)),
            interval_ms,
        }
    }

    pub fn start(&self) -> mpsc::Receiver<Tick> {
        let (tx, rx) = mpsc::channel(64);
        let symbol = self.symbol.clone();
        let price = self.price.clone();
        let interval_ms = self.interval_ms;

        tokio::spawn(async move {
            let mut ticker = interval(Duration::from_millis(interval_ms));
            loop {
                ticker.tick().await;
                let new_price = {
                    // Random walk: price moves +/- 0.1% per tick.
                    let drift = (rand::thread_rng().gen::<f64>() - 0.5) * 0.002;
                    let mut p = price.lock().unwrap();
                    *p += drift * *p;
                    *p
                };
                if tx
                    .send(Tick {
                        symbol: symbol.clone(),
                        price: new_price,
                        ts: Instant::now(),
                    })
                    .await
                    .is_err()
                {
                    break; // receiver dropped
                }
            }
        });

        rx
    }

    /// Mock fill at current price. Zero slippage, zero latency.
    pub async fn submit_order(&self, order: Order) -> Fill {
        let fill_price = *self.price.lock().unwrap();
        Fill {
            order,
            fill_price,
            ts: Instant::now(),
        }
    }
}
